### PR TITLE
Add Fabric Networking API v1

### PR DESCRIPTION
**Pull request by liach**
This fabric networking API v1 is a low-level API (or more exactly, hooks) that covers basic custom-payload/plugin message packets offered by the vanilla Minecraft protocol.

Usage: Users are supposed to use stuff exposed in `ClientNetworking` or `ServerNetworking`, or utilities in `net.fabricmc.fabric.api.networking.v1.util` package.

Initial content based off FabLabsMC/networking-api-v1-draft@eb256236cdfa9cb3b0697ceb12a957c47582fec3.
Javadoc for draft available at https://fablabsmc.github.io/networking-api-v1-draft/

Will perform v0 deprecation/migration only after the new api is accepted, or this pr is going to be too large.

Supersedes #455. Issue coverage: #164, #187, #247, #345, #424

Thanks @i509VCB for previewing the draft before I submit this pull request.

Signed-off-by: liach <liach@users.noreply.github.com>

### CODE REVIEW

File: fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/mixin/networking/ClientConnectionMixin.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	@Shadow
+	public abstract void disconnect(Text disconnectReason);
+
+	@Inject(method = "<init>", at = @At("RETURN"))
+	private void networking$ctor(NetworkSide side, CallbackInfo ci) {
+		this.playChannels = Collections.newSetFromMap(new ConcurrentHashMap<>());
+	}
+
+	@Redirect(method = "exceptionCaught", remap = false, at = @At(value = "INVOKE", target = "Lnet/minecraft/network/ClientConnection;send(Lnet/minecraft/network/Packet;Lio/netty/util/concurrent/GenericFutureListener;)V"))
+	private void networking$resendOnExceptionCaught(ClientConnection self, Packet<?> packet, GenericFutureListener<? extends Future<? super Void>> listener) {
```
Comment (modmuss50) [inline fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/mixin/networking/ClientConnectionMixin.java:64]: Id drop the $ in method names.

**Comment by modmuss50** [comment]
There is a lot to take in here, ill find some time to give it a propper look.

File: fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/mixin/networking/ClientConnectionMixin.java (hunk 1 above)
Comment (liach) [reply fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/mixin/networking/ClientConnectionMixin.java:64]: So should I drop all `networking$` prefix? (and pr fabric mixin to generate prefixes from mod ids?)

File: fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/mixin/networking/access/CustomPayloadC2SPacketAccess.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.gen.Accessor;
+
+import net.minecraft.network.PacketByteBuf;
+import net.minecraft.network.packet.c2s.play.CustomPayloadC2SPacket;
+import net.minecraft.util.Identifier;
+
+@Mixin(CustomPayloadC2SPacket.class)
+public interface CustomPayloadC2SPacketAccess {
```
Comment (i509VCB) [inline fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/mixin/networking/access/CustomPayloadC2SPacketAccess.java:27]: Hmm, I've noticed the naming for accessors is a bit inconsistent in fabric api.

Hook, Accessor, Access?

I'd lean towards `Accessor` but this inconsistency should be brought up in another PR to resolve it.

Comment (liach) [reply fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/mixin/networking/access/CustomPayloadC2SPacketAccess.java:27]: imo access and accessor are both the same. the interface does give access to the original object.

**Comment by ghost** [comment]
This seems really cool, it would be awesome to pick it back up again. Also, it might help if we came to a consensus about the whole player login event issue because it seems like this PR will solve all of those needs, but #226 and #671 also exist in order to solve this issue.

Ideally we ask the authors of that issue and PR if this would suit their needs, and then close the two to consolidate effort here.

**Comment by i509VCB** [comment]
This has been superseded by #1081. For liach, the new pull request is on the fablabs repo, so you can access it still.
